meta_error.go

```diff
@@ -0,0 +1,66 @@
+package microerror
+
+import (
+	"encoding/json"
+)
+
+// MetaError is a predefined error structure which purpose is to act as
+// container for meta information associated to a specific error. The specific
+// error type matching can be used as usual. The usual error masking and cause
+// gathering can be used as usual. Using MetaError might look as follows. In the
+// beginning is a usual error defined, along with its matcher. This error is the
+// root cause once emitted during runtime.
+//
+//     var notEnoughWorkersError = microerror.MetaError{
+//         Desc: "The amount of requested guest cluster workers exceeds the available number of host cluster nodes.",
+//         Docs: "https://github.com/giantswarm/ops-recipes/blob/master/349-not-enough-workers.md",
+//         Kind: "notEnoughWorkersError",
+//     }
+//
+//     func IsNotEnoughWorkers(err error) bool {
+//         return microerror.Cause(err) == notEnoughWorkersError
+//     }
+//
+type MetaError struct {
+	Desc string `json:"desc"`
+	Docs string `json:"docs"`
+	Kind string `json:"kind"`
+}
+
+func (m *MetaError) Error() string {
+	return toStringCase(m.Kind)
```

Comment:
I think generating the error message based on its kind makes total sense since this aligns with what we defined in our fmt repo. It is now that humans do not have to deal with this detail but only with setting the kind, which is more simple because it should just always be the variable name.

Reply:
I think it would be good to print `Desc` in `Error()`. And if this is the case `Desc` content shouldn't be a full sentence. I think about more generic errors like `Kind: "notFound"`. Do you think it makes sense to always have a specific error, e.g. `stackOutputNotFound`?

Also I would prefer to set the `Desc` with `Maskf` especially for those generic errors. When we do it like that it's easier to align description with log messages.

Reply:
> And if this is the case `Desc` content shouldn't be a full sentence.

I think the cool thing about the description is that you can write full sentences to describe the error and its purpose in more detail. I saw this for instance in the Azure API where errors are well explaining what is going on. I think this is a good thing. So I am fine with returning the description in `Error` though.

> I think about more generic errors like `Kind: "notFound"`. Do you think it makes sense to always have a specific error, e.g. `stackOutputNotFound`?

Both works IMO. That should also just be up to the dev implementing them based on the situation and its requirements. 

> Also I would prefer to set the `Desc` with `Maskf` especially for those generic errors. When we do it like that it's easier to align description with log messages.

You can call `Maskf` multiple times. This needs way more thinking because the behaviour isn't really clear to me then. There could be something like `Descf` which works against the description string so we could also make it a bit more dynamic. The example I made in the first place with having information about the requested number of nodes and the available number of nodes would then be addressed to some extend.

Reply:
So I understsand idea is to have `Desc` with full sentences but we still can annotate with `Maskf` to give more details. That makes sense for me. Maybe it isn't necessary to print the `Desc` in `Error()` then. Let's see.

Re manipulating stuff: note `var e := Error{}` is a value. So every time you call `Maskf` with `e` you get the copy. So we don't need to be afraid about manipulating in multiple places when we provide comparison function.